cmd/auth-provider-gcp/app/getcredentials.go

```diff
@@ -19,55 +19,91 @@ package app
 import (
 	"encoding/json"
 	"fmt"
-	"github.com/spf13/cobra"
 	"io/ioutil"
 	"net/http"
 	"os"
 
+	"github.com/spf13/cobra"
+
 	utilnet "k8s.io/apimachinery/pkg/util/net"
 	"k8s.io/cloud-provider-gcp/cmd/auth-provider-gcp/provider"
 	credentialproviderapi "k8s.io/cloud-provider-gcp/pkg/apis/credentialprovider"
 	"k8s.io/cloud-provider-gcp/pkg/credentialconfig"
 	klog "k8s.io/klog/v2"
 )
 
-var (
-	authFlow string
-)
-
 const (
 	gcrAuthFlow             = "gcr"
 	dockerConfigAuthFlow    = "dockercfg"
 	dockerConfigURLAuthFlow = "dockercfg-url"
 )
 
+type CredentialOptions struct {
+	AuthFlow string
+}
+
+type AuthFlowFlagError struct {
+	flagValue string
+}
+
+func (a *AuthFlowFlagError) Error() string {
+	return fmt.Sprintf("invalid value %q for authFlow (must be one of %q, %q, or %q)", a.flagValue, gcrAuthFlow, dockerConfigAuthFlow, dockerConfigURLAuthFlow)
+}
+
+func (a *AuthFlowFlagError) Is(err error) bool {
+	_, ok := err.(*AuthFlowFlagError)
+	return ok
+}
+
+type AuthFlowTypeError struct {
+	requestedFlow string
+}
+
+func (p *AuthFlowTypeError) Error() string {
+	return fmt.Sprintf("unrecognized auth flow \"%s\"", p.requestedFlow)
+}
+
+func (p *AuthFlowTypeError) Is(err error) bool {
+	_, ok := err.(*AuthFlowTypeError)
```

Comment:
Also check flagValue is what is expected? I believe that is in the spirit of the documented behavior of `errors.Is`.

Reply:
Sorry, what I mean is `requestedFlow` here and `flagValue` for `AuthFlowFlagError`

Reply:
This would mean anyone using `errors.Is` would have to know/set `flagValue` to the appropriate value in order to get `true` back - this is straightforward for unit tests, but would it be too difficult to use for other purposes? On the other hand, `AuthFlowTypeError`/`AuthFlowFlagError` are unrecoverable in the scope of auth-provider-gcp, so maybe it doesn't matter that much.

Reply:
You're right. We don't want to check these values. My bad.